cohort-evaluator-spark/src/main/java/com/ibm/cohort/cql/terminology/HashedCode.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * (C) Copyright IBM Corp. 2021, 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.ibm.cohort.cql.terminology;
+
+import java.util.Objects;
+
+import org.opencds.cqf.cql.engine.runtime.Code;
+
+
+/**
+ * This class is a simple wrapper class providing equals and hashCode functions
+ * for the org.opencds.cqf.cql.engine.runtime.Code class
+ * This is used to enable Set lookups for codes within ValueSets, which will be
+ * more performant than iterating through a list of codes
+ *
+ */
+public class HashedCode {
```

Comment:
This looks to be a duplicate of the CodeKey object in cohort-model-datarow. Do we need both?

Reply:
I did look at the CodeKey object, but in org.opencds.cqf.cql.engine.elm.execution.InValueSetEvaluator.inValueSet(), around line 49, there is a branch that passes in a code object with only a code specified and no system. Also, in the R4RestFhirTerminologyProviderTest, there is a testcase inOperationHandlesNullSystem() which implies that we need to support compare checks against code with no system defined.

For this reason, I needed to create my own class, because CodeKey does equality checks where the code system must be either present and equal on both codes being compared, or null on both. It doesn't support it being set on the code in the ValueSet read in from the file system, but null on the one passed into the in() method.

Reply:
> For this reason, I needed to create my own class

It's definitely one option. Another option would be to use `CodeKey` and build up an index explicitly using a Map or something rather than obscuring the logic in the `equals` and `hashCode` methods.

Reply:
I spent a lot of time on this today. I don't think there is clarity between the spec and the FHIR reference implementation on how this should work. The FHIR R4 Terminology Provider behaves in a way that isn't actually supported by FHIR (e.g. not providing the system parameter in the $validate-code request). The Javascript engine behaves in a clear way and reasonable way, but I don't think it is strictly matched up with the spec. I wrote up my notes on this topic and posted the question in [Zulip](https://chat.fhir.org/#narrow/stream/179220-cql/topic/'String'.20in.20MyValueSet) to see what the community has to say about it.

Reply:
Per Bryn, the correct behavior is documented in the CQL reference section [here](https://cql.hl7.org/09-b-cqlreference.html#in-valueset). 

> For the String overload, if the given valueset contains a code with an equivalent code element, the result is true. Note that for this overload, because the code being tested cannot specify code system information, if the resolved value set contains codes from multiple code systems, a run-time error is thrown because the operation is ambiguous.

I think the code index should be Map<String,Set<String>> where the key is the code and the value is the set of distinct codesystems associated with that code. When the lookup occurs on a code with no system, the provider should throw an "ambiguous lookup" exception.

```java
Set<String> systems = codeIndex.get(code);
if( systems != null && systems.size() > 1) { 
  throw new IllegalArgumentException("Ambiguous code lookup of code foo under valueset bar"); 
}
```